### ISSUE TITLE
Difficulty importing large datasets

### ISSUE THREAD

**Issue by cderenburger** [codeforamerica/ohana-api#361]
I'm attempting to import a whole statewide database and I've run into a few issues.

1. Importing script fails without an error message if it aborted due to being out of memory or it otherwise did not complete.
2. Importing seems to get exponentially longer the larger the database import.
 * Smaller db - 640 organizations, 1250 locations, 2440 services / ~30 minutes to import
    * adding taxonomy and service areas / ~60 minutes to import
 * Large db - 5110 organizations, 9920 locations, 20,000 services / 36 hours and counting...

My VM is still seems to be working on importing, ruby is using between 2.5-4G memory, 99% cpu.  I had thought it might finish at about 24 hours, but now I'm not sure it will even complete.  Watching <code>top</code> shows the postgres process running only a moment every 5-10 minutes.   

I'm not sure what my next step needs to be.  I'm wondering about either needing to set up a vagrant 64-bit box or attempting to build and merge dumps like described at http://blog.fespinozacast.com/2014/01/04/merge-two-backups-of-a-postgresql-database/ .

Any thoughts or suggestions?

**Comment by monfresh** [comment]
I haven't tested with such a large dataset, but it makes sense that the larger the data, the longer it will take, especially if the script has to geocode addresses. The machine's or VM's capabilities will play a major role in the speed as well. I would try updating the Vagrant box to a 64-bit machine and assigning it as much memory as possible (at least 4GB). I'll update the ohana-api-dev-box to do the same.

What kind of machine are you running the script on? I'm assuming a Windows machine. Would you be able to try it on a Mac with at least 8GB of RAM? If not, would you be able to share the CSV files with me so I can try it on my Mac?

**Comment by cderenburger** [comment]
Last night I put together an x64 Vagrant machine.  Pc specs are at. https://github.com/codeforamerica/ohana-api-dev-box/issues/12

All the addresses I geocoded beforehand as Google has a 2500 limit per day.  

On the x32 I had upped my VM memory size to 5.5G as ruby crashed when set to 4G.  After having upped to 5.5G the x32 VM would only use up to 4.2G, and would never pass importing services.csv after 48 hours.

So far the x64 seems to be working. I set the memory for the VM to 6.6G, and I'v seen it use as much as 6.2G. This morning my import is finally past services.csv and is currently working on phones.csv (my last table). I expect it to finish at about 14 or 16 hours, *(so long as my ups keeps up with the power failures today)*

**Comment by monfresh** [comment]
Thanks for the details. Have you tried running the script directly on your Linux box (not inside a VM)?

I suspect the script might be loading all CSV contents in memory. I will look into it and see if it can be optimized.

**Comment by md5** [comment]
@monfresh The script loads the whole file into memory with `File.read`, then loads all the CSV entries into memory with `CSV.new(...).entries`, then creates an array of `Organization` objects before saving them all.

I briefly played with using `Enumerator::Lazy` to optimize parts of it when I saw this thread, but it's going to be quite a bit of work (but well worth it I'd say).

**Comment by md5** [comment]
Using `CSV.new(...).each` instead of `#entries` would be a start, and I got most of the `OrganizationImporter` spec working with that change, but I didn't go all the way through the code to make it fully lazy.

**Comment by monfresh** [comment]
Thanks for looking into this, Mike. If you have time to keep working on it and feel like it, that would be great. I'm not sure I have much time to devote to this at the moment.

Cory, please let us know whether your latest attempt was successful. If not, for a short term fix, I would either try to find more RAM, or split your CSV files into smaller chunks.

**Comment by cderenburger** [comment]
> find more RAM, or split your CSV files into smaller chunks.

I just upgraded my nachine to 8GB, DDR2 is old and pricey so I am limited unless I build a new machine.
What do I need to do to split the files?

**Comment by monfresh** [comment]
To split the files, I was thinking you could take your `organizations.csv` and save it as a new file with half (or a quarter) of the rows, then repeat with all the other CSVs, while making sure the foreign keys between the CSVs match each other. Alternatively, if you have a script that generated the CSVs in the first place, run it again but with only a subset of the data at a time.

**Comment by cderenburger** [comment]
Thanks @monfresh, I will try that as well and let you kniw how it goes.

**Comment by monfresh** [comment]
@cderenburger Could you try this branch with your smaller dataset and see if it improves the speed? https://github.com/codeforamerica/ohana-api/tree/361-csv-memory

If that doesn't work, I'm thinking of looking into https://github.com/tilo/smarter_csv or https://github.com/fiksu/rcsv

**Comment by md5** [comment]
I've also started working on some code here: https://github.com/appropriate/ohana-api/tree/import-incremental

**Comment by md5** [comment]
More work is needed to update `LocationImporter`, remove `File.read` calls from `FileChecker` and `AddressExtractor`. Probably some more tweaks as well.

**Comment by monfresh** [comment]
Do you think it might be easier to use [smarter_csv](https://github.com/tilo/smarter_csv)? It provides the ability to process the CSV in chunks. Also, how are you testing the improvements? Do you have a large dataset to test with?

**Comment by md5** [comment]
@monfresh I don't think the CSV processor is the issue, particularly since the chunking would have to be smart enough to work across files as you said in another comment. Using a faster or less memory-intensive CSV processor wouldn't hurt, but I think the problem here is mainly that multiple copies of the same data is being retained in memory (leading to high memory requirements and GC pressure).

I see two main things that should be done to optimize the processing of large CSV files:

1. Make sure that everything is processing files, CSV rows, and objects in a streaming fashion
2. Aiming for constant memory use by persisting things to the DB as soon as possible

The work I did in my branch is a start for achieving number 1 (the changes you made partially address that as well). I also started to address the second point, but I think the whole thing would need to be wrapped in a transaction as well to avoid committing partial results (or to only commit partial results in a controlled manner).

**Comment by md5** [comment]
BTW, the memory use won't be constant to the extent that errors are being stored in memory, but I'd probably address that by having a maximum error count after which all processing is aborted. The DB memory use may also not be constant due to the open transaction, but DBs are usually better at managing that memory pressure and putting things on disk temporarily than any code you or I would write.

**Comment by md5** [comment]
Also, I haven't tested the improvements directly. I'm just making sure the specs pass at this point and doing general cleanups based on things that I've already found loading the whole dataset into memory.

I did do some standalone tests to make sure that `CSV.new(...).lazy.each` is not leaking when reading from an infinite stream (using the `yes` command).

**Comment by md5** [comment]
@monfresh Sounds good to me :+1: 

You may still want to look at my code for how I was able to strip most of the code out of the `EntityImporter` subclasses. I think it would make the whole thing easier to improve in the future.

**Comment by md5** [comment]
Also, I haven't made any changes to `LocationImporter` or the related classes for locations and addresses yet, so a comparison is probably premature.

**Comment by cderenburger** [comment]
I'm having an issue with the merge in my dev branch
running script.import:

rake aborted!
NameError: uninitialized constant FileChecker::SmarterCSV
/media/sf_vagrant/ohana-api/lib/file_checker.rb:64:in `csv_entries'
/media/sf_vagrant/ohana-api/lib/file_checker.rb:26:in `empty?'
/media/sf_vagrant/ohana-api/lib/file_checker.rb:34:in `required_but_empty?'
/media/sf_vagrant/ohana-api/lib/file_checker.rb:38:in `missing_or_empty?'
/media/sf_vagrant/ohana-api/lib/file_checker.rb:9:in `validate'
/media/sf_vagrant/ohana-api/lib/entity_importer.rb:9:in `check_and_import_file'
/media/sf_vagrant/ohana-api/lib/tasks/import.rake:9:in `block (2 levels) in <top (required)>'
-e:1:in `<main>'
Tasks: TOP => import:all => import:organizations

I'm busy at work, but trying to figure out where my issue lies.

**Comment by monfresh** [comment]
@cderenburger You'll need to run `bundle install` to pick up the `smarter_csv` gem.

**Comment by cderenburger** [comment]
I did pick up the gem, but perhaps I need to clear them out and grab them all again.  Had a similar issue with puma earlier.

I cloned a new directory from the master repo to test.  https://github.com/codeforamerica/ohana-api/tree/361-csv-memory seems to work great.  Far faster import, took 17 minutes for an import that usually take 45-60minutes.  Memory usage is around 1.75G

**Comment by cderenburger** [comment]
Re-running the tests with clean master copy:
**master**: Time: 44m  Memory: ~1.61G max
**361-csv-memory**: 44m  Time:  Memory: ~1.57G max

When I get home I'll upload a copy of my sample data for you to test with.  It will require settings.yml as I have service areas set.

**Comment by monfresh** [comment]
@cderenburger Thanks!

@md5 I saw that DRYing up. It looks nice.

**Comment by cderenburger** [comment]
You can find my small data set at https://github.com/cderenburger/ohana-api/tree/smaller-dataset

I've been testing and retesting checking memory usage, each time the results are inconsistent.  If I have the Virtualbox Manager open resident memory is about 1.5G, virtual mem about 2.5G, with Virtualbox Manager closed resident memory is 150M, virtual mem is 3G.  Yet in both tests overall system memory usage tops out 3.9G.  I've been doing a <code>vagrant halt</code> and <code>vagrant up</code> between each test.

Previous test results were from watching <code>top</code> mem used.

**Comment by cderenburger** [comment]
I finally installed the all the dependencies to run outside of the Vagrant VM.  Running without the Vagrant machine saves nearly a Gig of memory and cut import time 39 minutes.

**Comment by monfresh** [comment]
Thanks for the data. Is that the one that is taking 44 minutes for you? On my 2012 Macbook Air, it takes 31 minutes. The script takes so long because of the various `touch`es that are required to populate the full-text search index. To speed up the import, which should only be a one-time task, you can do a search for `, touch: true` and remove them temporarily while you import. In addition, you can remove or comment out the `after_add` and `after_remove` calls in `app/models/service.rb`, as well as this line:
```ruby
after_save :update_location_status, if: :status_changed?
```
**With these changes, the script ran in 8 minutes.**

Once the script is done, you can then populate the search index by running `Location.find_each(&:touch)` from the Rails console (`rails c`). That takes an additional 3 minutes. At this point, your DB should be ready to export via `script/export`, at which point you'll have a Postgres `.dump` file which you can then use to restore the DB via `script/restore`, which will only take seconds.

Finally, you'll need to add the code that you removed back.

The easiest solution to this speed problem is to update the installation instructions based on what I mentioned above. Off the top of my head, I can't think of an easy way to programmatically bypass the `touch`es during import.

**Comment by md5** [comment]
It looks like they're working on making the `touch` overhead less for Rails 5, but there doesn't appear to be a built-in solution for Rails 4. Here's the relevant PR: https://github.com/rails/rails/pull/19324

In the original issue (https://github.com/rails/rails/issues/18606, *Limit touching to once per transaction*), this gem is mentioned: https://github.com/godaddy/activerecord-delay_touching/

I believe that both the built-in solution for Rails 5 and the `activerecord-delay_touching` gem expect or possibly require the whole thing to be wrapped in a transaction. I suspect that wrapping it all in a transaction may also provide a performance gain all on its own, though you shouldn't expect more than a modest improvement.

**Comment by cderenburger** [comment]
>  Is that the one that is taking 44 minutes for you?

Outside of the VM it now takes 38 minutes for me to import.  Memory usage between the <code>master</code> and <code>361-csv-memory</code> branch seem to take the same amount of time to import and seemed to use the same amount of memory.  I just now saw your latest commit, I haven't merged it in yet to test, will try shortly.

In Vagrant
>**master:** Time: 44m Memory: ~1.62G max
**361-csv-memory:** 44m Time: Memory: ~1.57G max

Most recent test: 
**361-csv-memory:**  Time:  48m  Memory: ~2.1G
*each test in the VM had very inconsistent memory usage from one test to the next*

Outside Vagrant VM
**master:** Time: 38m  Memory: 1G
**361-csv-memory:**  Time 38m  Memory: 1G

Sorry all, seems that many of the issues I've been having are apparently related to Vagrant and VirtualBox.  I thought it was great to have a dev box I could build and destroy as needed, but the memory usage ended up being an issue.  When I start a test the VM uses ~500MB more, which is expected, but as the testing goes it balloons to needing 1G more memory than the same test outside the VM.  This doubled my memory needs for this smaller data set.  I'll have to see what the effect is on the large data set.

**Comment by monfresh** [comment]
It turns out Rails has a `no_touching` method, and wrapping all the `save` calls in a `ActiveRecord::Base.no_touching do` block does the trick. The import now takes less than 11 minutes total.

I've updated master with this latest code. Please pull it down again and give it a try. You won't need to change any code manually as I had mentioned in my previous comment.

**Comment by md5** [comment]
It seems like it might be good to still `touch` all the affected objects at the end of an import (though that may lead to cascading `touch`es through the `belongs_to` relations). Keeping track of the affected ids and `arel` instead of ActiveRecord to do a bulk update of the `updated_at` column should be pretty fast though.

**Comment by monfresh** [comment]
Yep. That's what [this](https://github.com/codeforamerica/ohana-api/blob/master/lib/tasks/import.rake#L70-L74) does, although it's not fast because each location's `tsv_body

**Comment by md5** [comment]
:+1:

**Comment by md5** [comment]
My one concern here would be that a partial import will still update the `updated_at` column of all locations, not just the ones that were imported.

Thanks for pointing me to the right part of the diff.

**Comment by cderenburger** [comment]
**Fantastic**
I finally got a chance to pull the latest <code>master</code> and the import with the smaller data set took me 15 minutes to import.  This is so much faster now.

Last night I had attempted to load my large data set.  It took about 45 minutes to load the data, then ran for the next 3 hours updating the full-text search.  I after the import script had gone to the "Updating full-text-search" I attempted to cancel the import.  I could not abort the process using <code>Ctrl + c</code>.  Later in the night it must have aborted at some point.  This morning I was unable to <code>script/reset</code>, I ended up having to <code>sudo -u postgres dropdb ohana_api_development</code> to finally drop the database.  Probably user error on my part.

I'll attempt the full import again today.  If the current master dropped my import time on the smaller data set to 1/3 the time, perhaps this will take about 8 hours.

I really appreciate the time and effort you have spent to help me on this issue.  Thanks @monfresh and @md5!
